src/goto-programs/graphml_witness.cpp

```diff
@@ -169,22 +272,93 @@ static bool filter_out(
      source_location.get_file().empty() ||
      source_location.is_built_in() ||
      source_location.get_line().empty())
+  {
+    const irep_idt id = source_location.get_function();
+    // Do not filter out assertions in functions the name of which starts with
+    // CPROVER_PREFIX, because we need to maintain those as violation nodes:
+    // these are assertions generated, for examples, for memory leaks.
+    if(!has_prefix(id2string(id), CPROVER_PREFIX) || !it->is_assert())
+      return true;
+  }
+
+  return false;
+}
+
+static bool contains_symbol_prefix(const exprt &expr, const std::string &prefix)
+{
+  if(
+    expr.id() == ID_symbol &&
+    has_prefix(id2string(to_symbol_expr(expr).get_identifier()), prefix))
+  {
     return true;
+  }
 
+  forall_operands(it, expr)
+  {
+    if(contains_symbol_prefix(*it, prefix))
+      return true;
+  }
   return false;
 }
 
 /// counterexample witness
 void graphml_witnesst::operator()(const goto_tracet &goto_trace)
 {
+  unsigned int max_thread_idx = 0;
+  bool trace_has_violation = false;
+  for(goto_tracet::stepst::const_iterator it = goto_trace.steps.begin();
+      it != goto_trace.steps.end();
+      ++it)
+  {
+    if(it->thread_nr > max_thread_idx)
+      max_thread_idx = it->thread_nr;
+    if(it->is_assert() && !it->cond_value)
+      trace_has_violation = true;
+  }
+
   graphml.key_values["sourcecodelang"]="C";
 
   const graphmlt::node_indext sink=graphml.add_node();
   graphml[sink].node_name="sink";
-  graphml[sink].thread_nr=0;
   graphml[sink].is_violation=false;
   graphml[sink].has_invariant=false;
 
+  if(max_thread_idx > 0 && trace_has_violation)
+  {
+    std::vector<graphmlt::node_indext> nodes;
+
+    for(unsigned int i = 0; i <= max_thread_idx + 1; ++i)
+    {
+      nodes.push_back(graphml.add_node());
+      graphml[nodes.back()].node_name = "N" + std::to_string(i);
+      graphml[nodes.back()].is_violation = i == max_thread_idx + 1;
+      graphml[nodes.back()].has_invariant = false;
+    }
+
+    for(auto it = nodes.cbegin(); std::next(it) != nodes.cend(); ++it)
+    {
+      xmlt edge("edge");
+      edge.set_attribute("source", graphml[*it].node_name);
+      edge.set_attribute("target", graphml[*std::next(it)].node_name);
+      const auto thread_id = std::distance(nodes.cbegin(), it);
+      xmlt &data = edge.new_element("data");
+      data.set_attribute("key", "createThread");
+      data.data = std::to_string(thread_id);
+      if(thread_id == 0)
+      {
+        xmlt &data = edge.new_element("data");
+        data.set_attribute("key", "enterFunction");
+        data.data = "main";
+      }
+      graphml[*std::next(it)].in[*it].xml_node = edge;
+      graphml[*it].out[*std::next(it)].xml_node = edge;
+    }
+
+    // we do not provide any further details as CPAchecker does not seem to
```

Comment:
Perhaps link to some appropriate issue or discussion thread?

Reply:
A link to furious discussions at the SV-COMP meeting in Thessaloniki? :-)  Or maybe https://github.com/sosy-lab/sv-witnesses/issues/12?